README.md

```diff
@@ -5,3 +5,134 @@
+1. The code required to turn `θ` into `model` inside `build_model` mustn't be too onerous to
+	write, read, or modify.
+
+While the first point is fairly straightforward, the second point is a bit subtle, so it's
+worth dwelling on it a little.
+
+For the sake of concreteness, let's suppose that we adopt the convention that `θ` is a
+`Vector{Float64}`. In the case of linear regression, we might assume that `θ` comprises
+a length `D` "weight vector" `w`, and a scalar "bias" `b`. So the function to build the
+model might be something like
+
+```julia
+function build_model(θ::Vector{Float64})
+    return x -> dot(θ[1:end-1], x) + θ[end]
+end
+```
+
+The easiest way to see that this is a less than ideal solution is to consider what this
+function would look like if `θ` was, say, a `NamedTuple` with fields `w` and `b`:
+```julia
+function build_model(θ::NamedTuple)
+    return x -> dot(θ.w, x) + θ.b
+end
+```
+This version of the function is much easier to read -- moreover if you want to inspect the
+values of `w` and `b` at some other point in time, you don't need to know precisely how to
+chop up the vector.
+
+Moreover it seems probable that the latter approach is less
+bug-prone -- suppose for some reason one refactored the code so that the first element of
+`θ` became `b` and the last `D` elements `w`; any code that depended upon the original
+ordering will now be incorrect and likely fail silently. The `NamedTuple` approach simply
+doesn't have this issue.
+
+Granted, in this simple case it's not too much of a problem, but it's easy to find
+situations in which things become considerably more difficult. For example, suppose that we
+instead had pretty much any kind of neural network, Gaussian process, ODE, or really just
+any model with more than a couple of distinct parameters. From the perspective of
+writing complicated models, implementing things in terms of a single vector of
+parameters that is _manually_ chopped up is an _extremely_ bad design choice. It simply
+doesn't scale.
+
+However, a single vector of e.g. `Float64`s _is_ extremely convenient when writing general
+purpose optimisers / approximate inference routines --
+[Optim.jl](https://github.com/JuliaNLSolvers/Optim.jl) and
+[AdvancedHMC.jl](https://github.com/TuringLang/AdvancedHMC.jl) being two obvious examples.
+
+
+
+
+
+# The ParameterHandling.jl Approach
+
+`ParameterHandling.jl` aims to give you the best of both worlds by providing the tools
+required to automate the transformation between a "structured" representation (e.g. nested
+`NamedTuple` / `Dict` etc) and a "flattened" (e.g. `Vector{Float64}`) of your model
+parameters.
+
+The function `flatten` eats a structured representation of some parameters, returning the
+flattened representation _and_ a function that converts the flattened thing back into its
+structured representation.
+
+`flatten` is implemented recursively, with a _very_ small number of base-implementations
+that don't themselves call `flatten`.
+
```

Comment:
Is `flatten` recursive so that it can flatten a multi dimensional structured representation?

Reply:
Exactly. For example, to `flatten` a `Tuple` you need to be able to `flatten` each of its elements and concatenate them.